[THREESCALE-2338] Performance documentation (#1068)
3scale/APIcast: doc/performance.md

# Performance

- [**General guidelines**](#general-guidelines)
- [**Default caching**](#default-caching)
- [**Async reporting threads**](#async-reporting-threads)
- [**Batching policy**](#batching-policy)

> This document gives some general guidelines to debug performance issues in

Reviewer: ```suggestion
This document provides general guidelines to debug performance issues in
```

> APIcast. It also introduces the caching modes available and explains how they

Reviewer: ```suggestion
APIcast. It also introduces the available caching modes and explains how they
```

can help in increasing performance.

## General guidelines

When experiencing performance issues in APIcast, the first thing that we

> recommend to do is to try to determine the component that is responsible for

Reviewer: ```suggestion
recommend to do is identifying the component responsible for
```

Reviewer 2: ```suggestion
recommend is to identifying the component that is responsible for
```

those issues. In a typical APIcast deployment, there are three components to

> take into account:

Reviewer: ```suggestion
consider:
```

- APIcast
- The 3scale backend that authorizes requests and keeps track of the usage
- The upstream API

> In order to measure the latency that APIcast plus the 3scale backend introduce,

Reviewer: ```suggestion
To measure the latency that APIcast plus the 3scale backend introduce,
```

the first thing that we need to do is to measure the latency of the upstream

> API. There are many tools that can be used for that:

Reviewer: In general, we avoid linking to third-party products in the main body of the docs. If mentioning these products is essential, you can create a separate section.

Author: I think it's fine. Those are open-source projects. In a product documentation I can see how this could become confusing regarding the level of support and other things, but in an upstream repo, I think it should be OK.

Reviewer: It's a matter of maintainability. If we have an "additional resources" section, we can easily locate and update links and any other content related to third-party components.
On the other hand, given the current state of technology, open source projects can change their licensing features, requiring us to quickly make necessary updates to avoid legal issues.

[wrk](https://github.com/wg/wrk), [vegeta](https://github.com/tsenart/vegeta),
[JMeter](https://jmeter.apache.org/), etc. The next step consists of using the
same tool to run the same benchmark, but this time, pointing to APIcast instead
of pointing to the upstream API directly. Comparing those results will give us
an idea of the latency introduced by APIcast and the 3scale backend.

> If you see that the latency introduced by APIcast and the 3scale backend is too

Reviewer: Reorganize the ideas from "If..."until "the latency". Suggestion:
The first thing you can do is to make a simple request to the 3scale backend from the same machine where APIcast is deployed, and measure the latency. This is possible given these requisites:
- If the latency introduced by APIcast and the 3scale backend is too high.
- You are using a self-managed APIcast that authorizes against the 3scale backend SaaS.

high, and you are using a self-managed APIcast that authorizes against the
3scale backend SaaS, the first thing you can do is to make a simple request to
the 3scale backend from the same machine where APIcast is deployed and measure

> the latency. Backend exposes a simple endpoint that just returns the version:

Reviewer: ```suggestion
the latency. Backend exposes an endpoint that returns the version:
```

> the latency. Backend exposes a simple endpoint that just returns the version:

Reviewer: Start a new paragraph from "Backend..."

> https://su1.3scale.net/status. An authorization call requires more work because

Reviewer: ```suggestion
https://su1.3scale.net/status. An authorization call requires 
```

Author: Removing that part would leave the sentence incomplete.

Reviewer: _An authorization call requires more work because the 3scale backend from the same machine where APIcast is deployed and measure the latency_ still is incomplete.
What did you want to state?

> it needs to verify some keys, limits, and enqueue background jobs, among other

Reviewer: ```suggestion
verifying some keys and limits, enqueuing background jobs, as well as ...
```
_among other things_ looks like a filler. If there are more tasks, specify them. Otherwise, you can finish the sentence.

Author: My intention was to mention only the features that are relevant for APIcast, and at the same time, don't give the impression that those are the only things that backend does.

Reviewer: Accuracy is paramount in technical documentation. We don't want to leave gaps that might cause confusion. 
I understand that this section refers mainly to APIcast. Better if you mention that these features are relevant for APIcast. Regarding the backend's functionality, you can mention a couple of tasks, or link to another doc for further reference.

things. The 3scale backend is very fast, on average, it does that in a matter of

> a few milliseconds, but still, it's more work than just checking the version

Reviewer: ```suggestion
a few milliseconds, but still, it requires more work than checking the version
```

> like the `/status` endpoint does. So, if a request to `/status` takes, let's say

Reviewer: ```suggestion
like the `/status` endpoint does. So, if a request to `/status` takes, as an example,
```

around 300 ms from your APIcast environment, an authorization is going to take a
bit more than that for every request that is not cached.

## Default caching

> For requests that are not cached, APIcast extracts the usage metrics from

Reviewer: Perhaps you want to reorganize this paragraph into a list of steps:

1. APIcast extracts the usage metrics from matching mapping rules
2. APIcast sends those metrics plus the application credentials
to the 3scale backend.
3. The 3scale backend makes the following tasks:
3.1.  Checks the application keys, and that the reported usage of metrics is within the defined limits.
3.2.  Enqueues a background job to increase the usage of the metrics reported.
3.3. Responds to APIcast whether the request should be authorized or not.
If it should be, the request goes to the upstream. Note that in this case, the request does not arrive to the upstream until the 3scale backend responds.

matching mapping rules and sends those metrics plus the application credentials
to the 3scale backend. The 3scale backend checks the application keys and also
that the usage of the metrics reported is within the defined limits. It also
enqueues a background job to increase the usage of the metrics reported. After
that, backend responds to APIcast whether the request should be authorized or
not. If it should be, the request goes to the upstream. Notice that in this
case, the request does not arrive to the upstream until the 3scale backend
responds.

> With the caching mechanism that comes enabled by default, APIcast stores in a

Reviewer: Add the steps of this paragraph to the previous suggested list.

cache the result of the authorization call to the 3scale backend if it was
authorized. The next request with the same credentials and metrics will use that
cached authorization instead of going to the 3scale backend. If the request was
not authorized or it is the first time that APIcast receives those credentials,
APIcast will call backend synchronously as explained above.

When the auth is cached, APIcast first calls the upstream and then, in a phase
called "post action", it calls the 3scale backend and stores the authorization
in the cache to have it ready for the next request. Notice that in this case,
the call to the 3scale backend does not introduce any latency because it does

> not happen in request time. There's an important aspect to take into account

Reviewer: ```suggestion
not happen in request time. There is an important aspect to take into account
```
